File: backend/internal/database/postgres.go
package database

import (
	"context"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
)

func Connect(url string) *pgxpool.Pool {
	pool, err := pgxpool.New(context.Background(), url)
	if err != nil {
		log.Fatalf("Unable to connect to database: %v", err)
	}
	return pool
}

func migrationDir() string {
	candidates := []string{"migrations", filepath.Join("backend", "migrations")}
	for _, dir := range candidates {
		if info, err := os.Stat(dir); err == nil && info.IsDir() {
			return dir
		}
	}
	log.Fatal("migrations directory not found (run from backend/ or project root)")
	return ""
}

func ensureMigrationTable(ctx context.Context, pool *pgxpool.Pool) {
	_, err := pool.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS schema_migrations (
			filename   VARCHAR(255) PRIMARY KEY,
			applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
		)`)
	if err != nil {
		log.Fatalf("Create schema_migrations: %v", err)
	}
}

func isMigrationApplied(ctx context.Context, pool *pgxpool.Pool, name string) bool {
	var exists bool
	err := pool.QueryRow(ctx, `SELECT EXISTS(SELECT 1 FROM schema_migrations WHERE filename = $1)`, name).Scan(&exists)
	return err == nil && exists
}

func markMigrationApplied(ctx context.Context, pool *pgxpool.Pool, name string) {
	_, err := pool.Exec(ctx, `INSERT INTO schema_migrations (filename) VALUES ($1) ON CONFLICT DO NOTHING`, name)
	if err != nil {
		log.Fatalf("Record migration %s: %v", name, err)
	}
}

func Migrate(pool *pgxpool.Pool) {
	dir := migrationDir()
	entries, err := os.ReadDir(dir)
	if err != nil {
		log.Fatalf("Read migrations: %v", err)
	}

	names := make([]string, 0)
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".sql") {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)

	ctx := context.Background()
	ensureMigrationTable(ctx, pool)

	for _, name := range names {
		if isMigrationApplied(ctx, pool, name) {
			log.Printf("⊘ Migration skipped (already applied): %s", name)
			continue
		}

		path := filepath.Join(dir, name)
		sqlBytes, err := os.ReadFile(path)
		if err != nil {
			log.Fatalf("Read migration %s: %v", name, err)
		}
		if _, err := pool.Exec(ctx, string(sqlBytes)); err != nil {
			log.Fatalf("Migration %s failed: %v", name, err)
		}
		markMigrationApplied(ctx, pool, name)
		log.Printf("✓ Migration applied: %s", name)
	}
	log.Println("✓ Database migrate check complete")
}

File: backend/src/shared/database/postgres.ts
import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';

let pool: Pool | null = null;

export function connect(databaseUrl: string): Pool {
  if (!databaseUrl) {
    console.error('DATABASE_URL is required');
    process.exit(1);
  }
  pool = new Pool({ connectionString: databaseUrl });
  return pool;
}

export function getPool(): Pool {
  if (!pool) throw new Error('Database not initialized');
  return pool;
}

function migrationDir(): string {
  const candidates = [
    path.join(process.cwd(), 'migrations'),
    path.join(process.cwd(), 'backend', 'migrations'),
  ];
  for (const dir of candidates) {
    if (fs.existsSync(dir) && fs.statSync(dir).isDirectory()) return dir;
  }
  console.error('migrations directory not found (run from backend/ or project root)');
  process.exit(1);
}

async function ensureMigrationTable(client: Pool): Promise<void> {
  await client.query(`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      filename   VARCHAR(255) PRIMARY KEY,
      applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
    )`);
}

async function isMigrationApplied(client: Pool, name: string): Promise<boolean> {
  const res = await client.query(
    'SELECT EXISTS(SELECT 1 FROM schema_migrations WHERE filename = $1)',
    [name],
  );
  return res.rows[0]?.exists === true;
}

async function markMigrationApplied(client: Pool, name: string): Promise<void> {
  await client.query(
    'INSERT INTO schema_migrations (filename) VALUES ($1) ON CONFLICT DO NOTHING',
    [name],
  );
}

export async function migrate(client: Pool): Promise<void> {
  const dir = migrationDir();
  const names = fs
    .readdirSync(dir)
    .filter((f) => f.endsWith('.sql'))
    .sort();

  await ensureMigrationTable(client);

  for (const name of names) {
    if (await isMigrationApplied(client, name)) {
      console.log(`⊘ Migration skipped (already applied): ${name}`);
      continue;
    }
    const sql = fs.readFileSync(path.join(dir, name), 'utf8');
    await client.query(sql);
    await markMigrationApplied(client, name);
    console.log(`✓ Migration applied: ${name}`);
  }
  console.log('✓ Database migrate check complete');
}
